File: Java/leetcode_1552_MagneticForceBetweenTwoBalls.java
class Solution {
    public Boolean canPlaceBalls(int x, int[] positions, int m){
        int ballsPlaced=1;
        int prevBallPos = positions[0];
        for(int i=0;i<positions.length && ballsPlaced <= m;i++)
        {
            if(positions[i]-prevBallPos>=x)
            {
                ballsPlaced++;
                prevBallPos = positions[i];
            }
            if(ballsPlaced>=m)
            {
                return true;
            }
        }
        return false;
    }
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int low = 1;
        int n = position.length;
        int high = (position[n-1]-position[0])/(m-1);
        int answer = 1;
        while(low<=high)
        {
            int mid = low+(high-low)/2;
            if(canPlaceBalls(mid, position, m))
            {
                answer = mid;
                low=mid+1;
            }
            else
            {
                high = mid-1;
            }
        }
        return answer;
    }
}

File: Scala/leetcode_1552_MagneticForceBetweenTwoBalls.scala
object Solution {
    def maxDistance(position: Array[Int], m: Int): Int = {
        var sortedPos = position.sorted
        var low: Int = 1
        val n: Int= sortedPos.length
        var high: Int = (sortedPos(n-1) - sortedPos(0))/ (m-1)
        var answer: Int = 0

        while (low <= high) {
            var mid: Int = low + (high - low)/2
            if (canPlaceBalls(mid, sortedPos, m)){
                answer = mid
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        answer
    }
    def canPlaceBalls(x: Int, positions: Array[Int], m: Int): Boolean = {
        var ballsPlaced: Int = 1
        var prevBall: Int = positions(0)

        for (p <- 1 until positions.length){
            if ((positions(p) - prevBall) >= x) {
                ballsPlaced += 1
                prevBall = positions(p)
            }
        }
        ballsPlaced >= m
    }
}